src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs

```diff
@@ -0,0 +1,444 @@
+                    if ((uint) (l - 0xdd22) <= (uint) (0xdd43 - 0xdd22))
+                    {
+                        hr = h;
+                        lr = (ushort) ((l - 0xdd22) +  0xdd00);
+                        return;
+                    }
+                    break;
+            }
+
+            hr = h;
+            lr = l;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool EqualSurrogate(char h1, char l1, char h2, char l2)
+        {
+            ToUpperSurrogate(h1, l1, out ushort hr1, out ushort lr1);
+            ToUpperSurrogate(h2, l2, out ushort hr2, out ushort lr2);
+
+            return hr1 == hr2 && lr1 == lr2;
+        }
+
+        internal static int CompareStringIgnoreCase(ref char strA, int lengthA, ref char strB, int lengthB)
+        {
+            Debug.Assert(!GlobalizationMode.Invariant);
+            Debug.Assert(!GlobalizationMode.UseNls);
+
+            int length = Math.Min(lengthA, lengthB);
+
+            ref char charA = ref strA;
+            ref char charB = ref strB;
+
+            while (length != 0)
+            {
+                 // optimize for Ascii cases
+                if (charA <= '\u00FF' || length == 1 || !char.IsHighSurrogate(charA) || !char.IsHighSurrogate(charB))
+                {
+                    if (charA == charB)
+                    {
+                        length--;
+                        charA = ref Unsafe.Add(ref charA, 1);
+                        charB = ref Unsafe.Add(ref charB, 1);
+                        continue;
+                    }
+
+                    char aUpper = OrdinalCasing.ToUpper(charA);
+                    char bUpper = OrdinalCasing.ToUpper(charB);
+
+                    if (aUpper == bUpper)
+                    {
+                        length--;
+                        charA = ref Unsafe.Add(ref charA, 1);
+                        charB = ref Unsafe.Add(ref charB, 1);
+                        continue;
+                    }
+
+                    return aUpper - bUpper;
+                }
+
+                // We come here only of we have valid high surrogates and length > 1
+
+                char a = charA;
+                char b = charB;
+
+                length--;
+                charA = ref Unsafe.Add(ref charA, 1);
+                charB = ref Unsafe.Add(ref charB, 1);
+
+                if (!char.IsLowSurrogate(charA) || !char.IsLowSurrogate(charB))
+                {
+                    // malformed Surrogates - should be rare cases
+                    if (a != b)
+                    {
+                        return a - b;
+                    }
+
+                    // Should be pointing to the right characters in the string to resume at.
+                    // Just in case we could be pointing at high surrogate now.
+                    continue;
+                }
+
+                // we come here only if we have valid full surrogates
```

Comment:
Nit: This comment isn't 100% correct. It's possible for the contents of the `ROS<char>` input to have changed between the first read (line 235) and the second read (line 261), which makes this no longer a valid surrogate pair. The `ToUpperSurrogate` appears is resilient to malformed surrogate pairs, but I wouldn't make such a solid "this is valid" statement in a comment.

(This applies to a few other places in this file, such as the pointer-based routine below.)

Reply:
WOW. I like the security thinking here :-) if someone change the buffer contents during the operation they already missed up anyway. I wouldn't care much about that. We may clarify the comment more later but I would avoid that as it may suggest allow changing the underlying buffer.